lib/engine/step/special_train.rb

```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+require_relative 'base'
+require_relative 'trainer'
+
+module Engine
+  module Step
+    class SpecialTrain < Base
+      include Trainer
+
+      ACTIONS = %w[buy_train].freeze
+
+      def actions(entity)
+        return [] unless ability(entity)
+
+        ACTIONS
+      end
+
+      def blocks?
+        false
+      end
+
+      def process_buy_train(action)
+        buy_train_action(action, action.entity.owner)
+        ability = ability(action.entity)
```

Comment:
This has a problem in EMR that you can sell more shares than needed to use the discount, probably can be done in a follow up PR

Reply:
@jenf Thinking about it I am sceptical that company abilities could be used by the president to affect his/her personal resources. So I do not think the train discount ability would be applicable during an EMF.
Anyway, I have asked a question about it on BGG.

I suggest that we do not support that. I will do a test to see that this special train are not triggered by an emergency buy.

Reply:
I think we need to support using NDY during EMF, but as you say, it can be done in a future PR.